Sources/YUL/AST.swift

```diff
@@ -0,0 +1,342 @@
+//
+//  AST.swift
+//  YUL
+//
+//  Created by Yicheng Luo on 11/16/18.
+//
+
+import Foundation
+
+public typealias Identifier = String
+
+public enum Literal: CustomStringConvertible {
+  case num(Int)
+  case string(String)
+  case bool(Bool)
+  case decimal(Int, Int)
+  case hex(String)
+
+  public var description: String {
+    switch self {
+    case .num(let i):
+      return String(i)
+    case .string(let s):
+      return "\"\(s)\""
+    case .bool(let b):
+      return b ? "1" : "0"
+    case .decimal(let (n1, n2)):
+      return "\(n1).\(n2)"
+    case .hex(let h):
+      return h
+    }
+  }
+}
+
+public struct Block: CustomStringConvertible {
+  public var statements: [Statement]
+
+  public init(_ statements: [Statement]) {
+    self.statements = statements
+  }
+
+  public var description: String {
+    let statement_description = self.statements.map({ s in
+      return s.description
+    }).joined(separator: "\n")
+
+    return """
+    {
+      \(statement_description.indented(by: 2))
+    }
+    """
+  }
+}
+
+// swiftlint:disable type_name
+public struct If: CustomStringConvertible {
+// swiftlint:enable type_name
+  public let expression: Expression
+  public let block: Block
+
+  public init(_ expression: Expression, _ block: Block) {
+    self.expression = expression
+    self.block = block
+  }
+
+  public var description: String {
+    return "if \(expression.description) \(self.block)"
+  }
+}
+
+public struct Switch: CustomStringConvertible {
+  public let expression: Expression
+  public let cases: [(Literal, Block)]
+  public let `default`: Block?
+
+  public init(_ expression: Expression, cases: [(Literal, Block)], `default`: Block? = nil) {
+    self.expression = expression
+    self.cases = cases
+    self.default = `default`
+  }
+
+  public init(_ expression: Expression, `default`: Block? = nil) {
+    self.init(expression, cases: [], default: `default`)
+  }
+
+  public var description: String {
+    let cases = self.cases.map { (lit, block) in
+      return "case \(lit) \(block)"
+      }.joined(separator: "\n")
+
+    let `default` = self.default != nil ? "\ndefault \(self.default!)" : ""
```

Comment:
I know it's ugly but it'd be more readable if you had a conditional binding and just initialize the variable separately for each of the cases.

```swift
let `default`: String
if let defaultContent = self.default {
  `default` = "\ndefault \(defaultContent)"
} else {
  `default` = ""
}
```